### PR TITLE
Impl `np.shape` and support function for `asarray`.

### LINKED ISSUE #4080: typing error numpy.shape "not supported"
```
using spyder 3.3.4, python 3.7.0 in anaconda environment 1.9.7 on OS X 10.14.4 (mojave)
get this:
TypingError: Use of unsupported NumPy function 'numpy.shape' or unsupported use of the function.
File "testcode.py", line 23:
def ggp_timestep(g, Pfr):
    <source elided>
    sg=np.shape(g)       #this is line 23 in testcode.py
    thearr = np.random.rand(sg[0], sg[1])
```

Come on guys, it says over and over in the reference manual that shape is a supported numpy function.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Fixes #4080

### CODE REVIEW

File: numba/tests/test_array_manipulation.py [hunk 1]
```diff
@@ -708,5 +712,30 @@ def _assert_raises(arr, val):
         _assert_raises(arr, val)
 
 
+    def test_shape(self):
+        pyfunc = numpy_shape
+        cfunc = jit(nopython=True)(pyfunc)
+
+        def check(x):
+            expected = pyfunc(x)
+            got = cfunc(x)
```
Comment (esc) [inline numba/tests/test_array_manipulation.py]: Can we rename `got` to `received`? 'get', 'got' and 'gotten' are fairly unspecific.

**Comment by esc** [comment]
This looks good. @stuartarchibald do you envision `type_can_asarray` to be re-used in different places too?

**Comment by esc** [comment]
The azure failure is a flake8 error.

**Comment by stuartarchibald** [comment]
@esc yes `type_can_asarray` is in `numpy_support` with view of being used elsewhere.

File: numba/tests/test_array_manipulation.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/tests/test_array_manipulation.py]: Whilst I appreciate that `got` and variants are not ideal, the continued use is consistent with the rest of the code base, therefore I'm inclined to leave it.

**Comment by stuartarchibald** [comment]
Thanks for the review, I think I've addressed your comments.

File: numba/tests/test_array_manipulation.py (hunk 1 above)
Comment (esc) [reply numba/tests/test_array_manipulation.py]: Check, consistency wins.

**Comment by esc** [comment]
@stuartarchibald sounds good, can be marked to be merged when CI passes.

**Comment by seibert** [comment]
CI has passed.  Marking this as ready to merge.

**Pull request merged**